### PR TITLE
[EME] Refactored the GStreamer/Decryption part of the EME implementat…

### PR DESCRIPTION
**Pull request by woutermeek**
…ion.

      Highlights:
      - OPENCDM flag now selects OpenCDM implementation. No other decryption is included. All querying is done on OpenCDM server level.
      - If the OPENCDM flag is *NOT* selected the ClearKey implementation is included only.
      - PLayReady removed as an option, OPENCDM should be used for this.

### CODE REVIEW

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 1]
```diff
@@ -149,24 +148,55 @@ static const GstStreamVolumeFormat volumeFormat = GST_STREAM_VOLUME_FORMAT_LINEA
...
+    ASSERT(streamEncryptionAllowedSystemsValue && G_VALUE_HOLDS(streamEncryptionAllowedSystemsValue, G_TYPE_STRV));
+
+    const char** streamEncryptionAllowedSystems = reinterpret_cast<const char**>(g_value_get_boxed(streamEncryptionAllowedSystemsValue));
+
+    GST_TRACE("extractEventsAndSystemsFromMessage message: %" GST_PTR_FORMAT, message);
+
+    ASSERT(streamEncryptionAllowedSystems);
+    Vector<String> streamEncryptionAllowedSystemsVector;
+    unsigned i;
+    if (streamEncryptionAllowedSystems != NULL) {
```
Comment (charlie-ht) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Please fix the formatting here

File: Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp [hunk 2]
```diff
@@ -224,8 +223,8 @@ CDMInstanceOpenCDM::CDMInstanceOpenCDM(media::OpenCdm& system, const String& key
 {
     m_openCDM.SelectKeySystem(keySystem.utf8().data());
 
-    if (GStreamerEMEUtilities::isWidevineKeySystem(keySystem))
-        m_mimeType = "video/mp4";
+    /* if (GStreamerEMEUtilities::isWidevineKeySystem(keySystem))
```
Comment (charlie-ht) [inline Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp]: Please just remove

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 3]
```diff
@@ -93,8 +119,9 @@ static gboolean webKitMediaOpenCDMDecryptorHandleKeyResponse(WebKitMediaCommonEn
 {
     bool returnValue = false;
     const GstStructure* structure = gst_event_get_structure(event);
-    if (!gst_structure_has_name(structure, "drm-session"))
+    if (!gst_structure_has_name(structure, "drm-session")) {
```
Comment (charlie-ht) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: You don't need the braces

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 4]
```diff
@@ -105,11 +132,13 @@ static gboolean webKitMediaOpenCDMDecryptorHandleKeyResponse(WebKitMediaCommonEn
 
     GST_DEBUG_OBJECT(self, "handling session %s for event %u (ours %u)", session.get(), protectionEvent, priv->m_protectionEvent);
     if (priv->m_protectionEvent == protectionEvent) {
-        if (priv->m_session != session.get()) {
+        if (priv->m_openCdm == nullptr) {
```
Comment (charlie-ht) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Why the change in logic here?

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 5]
```diff
@@ -105,11 +132,13 @@ static gboolean webKitMediaOpenCDMDecryptorHandleKeyResponse(WebKitMediaCommonEn
...
             priv->m_session = session.get();
             priv->m_openCdm = std::make_unique<media::OpenCdm>(priv->m_session.utf8().data());
-            GST_DEBUG_OBJECT(self, "selected session %s", priv->m_session.utf8().data());
             returnValue = true;
+        }
+        else if (priv->m_session != session.get()) {
+            GST_DEBUG_OBJECT(self, "Different session send to:  %s - %s", priv->m_session.utf8().data(), session.get());
```
Comment (charlie-ht) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Would write this has "selected a new session %s, old one was %s" to make it easier to understand

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 4 above)
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: From a CDM/DRM Session perspective, it should not change during the playback (can change from key but not from session), The logic know is assuming this and logs a debug message (in the else) if this happens. 
It is assumed that programatically the Session is set by now. Potentially we might add an if with a warning for that.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 5 above)
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: It is not selecting a new session. It is reporting that a new session is now calling this method, compared to the one previously set. I think this should not happen. Might put an assert here.
I think changing the logic that if a new session is send, we might take it up as being the active session requires a more intrusive approach. Than we need to think about locking as the current session might be in use with a decrypt and we just replace it by a new one. The sessions should become than reference counted and it will, potentially, bring more locking consequences (and overhead) for something that I think, is form a design impossible.

**Comment by calvaris** [comment]
The intention of this is good and I could support something like this. Now the big issue: this does not work (unless I am very mistaken) with non-MSE playback. The UUIDs are something coming form the media and GStreamer relies on them to select the proper decryptor. Cabling the Widevine UUID will break this and racecar with PlayReady should not work.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 6]
```diff
@@ -406,7 +436,7 @@ bool MediaPlayerPrivateGStreamerBase::handleSyncMessage(GstMessage* message)
         }
         GST_DEBUG("handling drm-preferred-decryption-system-id need context message");
         LockHolder lock(m_protectionMutex);
-        std::pair<Vector<GRefPtr<GstEvent>>, Vector<String>> streamEncryptionInformation = GStreamerEMEUtilities::extractEventsAndSystemsFromMessage(message);
+        std::pair<Vector<GRefPtr<GstEvent>>, Vector<String>> streamEncryptionInformation = extractEventsAndSystemsFromMessage(message);
```
[the remark this answers is not in the document]
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: CAPS have been set properly no (added WideVine/Playready) Racecar is playing EME2018 tes page succeeds, so what can be done here? Really love to remove this as I never got what it was doing.

File: Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp [hunk 7]
```diff
@@ -99,9 +98,9 @@ std::unique_ptr<CDMPrivate> CDMFactoryOpenCDM::createCDM(const String& keySystem
     return std::unique_ptr<CDMPrivate>(new CDMPrivateOpenCDM(keySystem));
 }
 
-bool CDMFactoryOpenCDM::supportsKeySystem(const String& keySystem)
+bool CDMFactoryOpenCDM::supportsKeySystem(const String& /* keySystem */)
 {
-    return GStreamerEMEUtilities::isPlayReadyKeySystem(keySystem) || GStreamerEMEUtilities::isWidevineKeySystem(keySystem);
+    return true;
```
[the remark this answers is not in the document]
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp]: See last commit. Although I still think that, also prior to this change, the "pericodelospalotes.keysystem.com" would not lead to a DRM session since another call is made to IsTypeSupported that will return later on when the mime types are available. 
Now the call to IsTypeSupported from an OpenCDM level is done twice.

File: Source/WebCore/platform/GStreamer.cmake [hunk 8]
```diff
@@ -143,7 +141,6 @@ if (ENABLE_ENCRYPTED_MEDIA)
         ${LIBGCRYPT_INCLUDE_DIRS}
     )
     list(APPEND WebCore_SOURCES
-        platform/encryptedmedia/clearkey/CDMClearKey.cpp
```
Comment (blino) [inline Source/WebCore/platform/GStreamer.cmake]: Why do you disable ClearKey in the OpenCDM case?
It can be kept, and is still useful to run generic EME tests.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 9]
```diff
@@ -44,6 +45,28 @@ static gboolean webKitMediaOpenCDMDecryptorHandleKeyResponse(WebKitMediaCommonEn
...
+static GstStaticPadTemplate sinkTemplate = GST_STATIC_PAD_TEMPLATE("sink",
+    GST_PAD_SINK,
+    GST_PAD_ALWAYS,
+    GST_STATIC_CAPS(
+    "application/x-cenc, original-media-type=(string)video/webm, protection-system=(string)" WEBCORE_CDMFACTORY_SYSTEM_UUID "; "
+    "application/x-cenc, original-media-type=(string)video/mp4, protection-system=(string)" WEBCORE_CDMFACTORY_SYSTEM_UUID "; "
+    "application/x-cenc, original-media-type=(string)audio/webm, protection-system=(string)" WEBCORE_CDMFACTORY_SYSTEM_UUID "; "
+    "application/x-cenc, original-media-type=(string)audio/mp4, protection-system=(string)" WEBCORE_CDMFACTORY_SYSTEM_UUID "; "
+    "application/x-cenc, original-media-type=(string)video/x-h264, protection-system=(string)" WEBCORE_CDMFACTORY_SYSTEM_UUID "; "
+    "application/x-cenc, original-media-type=(string)audio/mpeg, protection-system=(string)" WEBCORE_CDMFACTORY_SYSTEM_UUID ";"));
```
[the remark this answers is not in the document]
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Added PlayReady and WideVine properties. non-MSE path is playing fine now.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp [hunk 10]
```diff
@@ -99,6 +98,45 @@ static void appendPipelineElementMessageCallback(GstBus*, GstMessage* message, A
...
+            GstStaticPadTemplate* staticPadTemplate = static_cast<GstStaticPadTemplate*>(current->data);
+            GRefPtr<GstCaps> caps = adoptGRef(gst_static_pad_template_get_caps(staticPadTemplate));
+            unsigned length = gst_caps_get_size(caps.get());
+
+            GST_TRACE("factory %s caps has size %u", GST_OBJECT_NAME(factory), length);
+            for (unsigned i = 0; !decryptor && i < length; ++i) {
+                GstStructure* structure = gst_caps_get_structure(caps.get(), i);
+                GST_TRACE("checking structure %s", gst_structure_get_name(structure));
+                if (gst_structure_has_field_typed(structure, GST_PROTECTION_SYSTEM_ID_CAPS_FIELD, G_TYPE_STRING)) {
+                    const char* sysId = gst_structure_get_string(structure, GST_PROTECTION_SYSTEM_ID_CAPS_FIELD);
```
[the remark this answers is not in the document]
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: Don understand what you mean here

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 11]
```diff
@@ -70,6 +95,7 @@ static void webkit_media_opencdm_decrypt_class_init(WebKitOpenCDMDecryptClass* k
     cencClass->handleKeyResponse = GST_DEBUG_FUNCPTR(webKitMediaOpenCDMDecryptorHandleKeyResponse);
     cencClass->decrypt = GST_DEBUG_FUNCPTR(webKitMediaOpenCDMDecryptorDecrypt);
     cencClass->receivedProtectionEvent = GST_DEBUG_FUNCPTR(webKitMediaOpenCDMDecryptorReceivedProtectionEvent);
+    cencClass->protectionSystemId = WEBCORE_CDMFACTORY_SYSTEM_UUID;
```
[the remark this answers is not in the document]
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Agree, but as you said this is how it was :-) More improvements could be made but they require interface changes and sofar I managed to stay within the interfaces as defined by....???? the openCDM/Gstreamer stuff

File: Source/WebCore/platform/GStreamer.cmake (hunk 8 above)
Comment (pwielders) [reply Source/WebCore/platform/GStreamer.cmake]: ClearKey is *NOT* disabled if you select OpenCDM. It is just running through the OpenCDM, to be handled in the OpenCDM server. If you deselect OpenCDM, the ClearKey is selected as the dafult DRM system, if you select OpenCDM, it is assumed that all DRM systems (including ClearKey) are running in the OpenCDM server)

Comment (blino) [reply Source/WebCore/platform/GStreamer.cmake]: I don't see the ClearKey keysystem in WebKitOpenCDMDecryptorGStreamer.cpp caps.

File: Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (charlie-ht) [reply Source/WebCore/platform/graphics/gstreamer/eme/CDMOpenCDM.cpp]: We discussed this change a while ago, we already do key system checks in the CDMPrivate's supportsConfiguration member call (the thing that is supposed to check supported mime / init data types). By returning false early here in the factory, all we avoid is creating a CDMPrivate, which is a tiny object that isn't worth sweating about. The m_openCdm member will now have to be added the CDMFactoryOpenCDM so it's still getting created. The only alternative I see to avoid doing a cross-process request for this is to hardcode what keysystems we support in opencdm into webkit, this is ugly, but practical. We don't add new keysystem support very often..

Anyway, that was just a dump of my recollection of the last conversation we had about this small point :-)

File: Source/WebCore/platform/GStreamer.cmake (hunk 8 above)
Comment (pwielders) [reply Source/WebCore/platform/GStreamer.cmake]: You just missed the last push :-)

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 12]
```diff
@@ -149,24 +148,55 @@ static const GstStreamVolumeFormat volumeFormat = GST_STREAM_VOLUME_FORMAT_LINEA
...
+        for (i = 0; streamEncryptionAllowedSystems[i]; ++i) {
+            streamEncryptionAllowedSystemsVector.append(streamEncryptionAllowedSystems[i]);
+        }
+    }
+
+    const GValue* streamEncryptionEventsList = gst_structure_get_value(structure, "stream-encryption-events");
+    ASSERT(streamEncryptionEventsList && GST_VALUE_HOLDS_LIST(streamEncryptionEventsList));
+    unsigned streamEncryptionEventsListSize = gst_value_list_get_size(streamEncryptionEventsList);
+    Vector<GRefPtr<GstEvent>> streamEncryptionEventsVector;
+    for (i = 0; i < streamEncryptionEventsListSize; ++i) {
```
[the remark this answers is not in the document]
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: I see this pending.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: You should be able to remove all included in the new line 428. Please consider the associated attributes that shouldn't be used anymore, like `m_reportedProtectionEvents`.

File: Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp [hunk 13]
```diff
@@ -149,24 +148,53 @@ static const GstStreamVolumeFormat volumeFormat = GST_STREAM_VOLUME_FORMAT_LINEA
...
+    ASSERT(streamEncryptionAllowedSystemsValue && G_VALUE_HOLDS(streamEncryptionAllowedSystemsValue, G_TYPE_STRV));
+
+    const char** streamEncryptionAllowedSystems = reinterpret_cast<const char**>(g_value_get_boxed(streamEncryptionAllowedSystemsValue));
+
+    GST_TRACE("extractEventsAndSystemsFromMessage message: %" GST_PTR_FORMAT, message);
+
+    ASSERT(streamEncryptionAllowedSystems);
+    Vector<String> streamEncryptionAllowedSystemsVector;
+    unsigned i;
+    if (streamEncryptionAllowedSystems != nullptr)
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamerBase.cpp]: Actually I made a mistake here, it should be `if (!streamEncryptionAllowedSystems)`. That's what the WebKit Coding Style says. Before I got trapped already with the `NULL`. That's why I say that a wrong style can make us lose more important issues if you're in a hurry.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (charlie-ht) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: @pwielders , he just wants the sysId variable name (`const char* sysId`) to be renamed to something clearer. In WebKit style it is encouraged to use full English words in the cases where abbreviations are not incredibly well known (sys for system is not acceptable, but Id for identification is), bit of a grey area... :)

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitClearKeyDecryptorGStreamer.h [hunk 14]
```diff
@@ -21,12 +21,14 @@
...
 #pragma once
 
-#if ENABLE(ENCRYPTED_MEDIA) && USE(GSTREAMER)
+#if ENABLE(ENCRYPTED_MEDIA) && USE(GSTREAMER) && !USE(OPENCDM)
 
 #include "WebKitCommonEncryptionDecryptorGStreamer.h"
 
 G_BEGIN_DECLS
 
+#define WEBCORE_CDMFACTORY_SYSTEM_UUID "58147ec8-0423-4659-92e6-f52c5ce8c3cc"
```
[the remark this answers is not in the document]
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitClearKeyDecryptorGStreamer.h]: Give me a suggestion for the name of this define and I will change it.

[the remark this answers is not in the document]
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitClearKeyDecryptorGStreamer.h]: This is still pending. If you remove the code I said this won't need to have the same name and that makes readability harder. A better name would be `WEBKIT_CLEARKEY_DECRYPTOR_GSTREAMER_UUID` and it doesn't need to be included in the headers anymore.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 15]
```diff
@@ -44,6 +49,49 @@ static gboolean webKitMediaOpenCDMDecryptorHandleKeyResponse(WebKitMediaCommonEn
 static gboolean webKitMediaOpenCDMDecryptorDecrypt(WebKitMediaCommonEncryptionDecrypt*, GstBuffer*, GstBuffer*, unsigned, GstBuffer*);
 static void webKitMediaOpenCDMDecryptorReceivedProtectionEvent(WebKitMediaCommonEncryptionDecrypt*, unsigned);
 
+static GstStaticPadTemplate sinkTemplate = GST_STATIC_PAD_TEMPLATE("sink",
+    GST_PAD_SINK,
+    GST_PAD_ALWAYS,
+    GST_STATIC_CAPS(
+    "application/x-cenc, original-media-type=(string)video/webm, protection-system=(string)" WEBCORE_CDMFACTORY_SYSTEM_UUID "; "
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: You don't need MSE and non-MSE UUIDs. The same should be used. These are the capabilities accepted by the decryptor and it needs to be able to accept just the three UUIDs once. The decryptor creation, both in GStreamer and in WebKit, should already be able to detect this.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 11 above)
[the remark this answers is not in the document]
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: The problem is that it is wrong and we need to fix this. I bet this is why you need the MSE and non-MSE caps.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 5 above)
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Regarding the semantics of session selection, the message should be formatted like @charlie-ht says.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp (hunk 10 above)
[the remark this answers is not in the document]
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: Variable names can be better and more understandable if:
* `sysId` -> `protectionSystemUuid`
* `protectionSystem` -> `requestedProtectionSystemUuid`

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp [hunk 16]
```diff
@@ -33,6 +34,10 @@
 
 #define GST_WEBKIT_OPENCDM_DECRYPT_GET_PRIVATE(obj) (G_TYPE_INSTANCE_GET_PRIVATE((obj), WEBKIT_TYPE_OPENCDM_DECRYPT, WebKitOpenCDMDecryptPrivate))
 
+#define WEBCORE_NON_MSE_PLAYREADY_UUID "9a04f079-9840-4286-ab92-e65be0885f95"
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: Better names for these would be `WEBKIT_OPENCDM_DECRYPTOR_GSTREAMER_*_UUID`

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 4 above)
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: I haven't checked the logic here, actually I am changing this in my branch.

Anyway, the style says it should be `if (priv->m_openCdm)`

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.h [hunk 17]
```diff
@@ -28,6 +28,8 @@
 
 G_BEGIN_DECLS
 
+#define WEBCORE_CDMFACTORY_SYSTEM_UUID "edef8ba9-79d6-dead-beef-27dcd51d21ed"
```
[the remark this answers is not in the document]
Comment (calvaris) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.h]: I don't think these need to be exposed in the headers anymore.

File: Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp (hunk 15 above)
Comment (pwielders) [reply Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp]: But the problem here is that I can only return one in the protection Id in the message. OpenCDM allows for, potentially, all CDMs. but than can not be returned from in the message. I suggest to do this in a next round.
So do agree that that should work, but this requires more modifications ( and I prefer to take it one step at a time.
https://github.com/WebPlatformForEmbedded/WPEWebKit/blob/pwielders/gstreamer-decryption-refactor/Source/WebCore/platform/graphics/gstreamer/eme/WebKitOpenCDMDecryptorGStreamer.cpp#L123
Here I can only fill in 1 protectionSystemId, in OpenCDM context this could be multiple and personally I don like to "abuse" one of the existing (PlayReady/WideVine/Clearkey) properties to use in this spot.

File: Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp [hunk 18]
```diff
@@ -99,6 +98,45 @@ static void appendPipelineElementMessageCallback(GstBus*, GstMessage* message, A
...
+            GstStaticPadTemplate* staticPadTemplate = static_cast<GstStaticPadTemplate*>(current->data);
+            GRefPtr<GstCaps> caps = adoptGRef(gst_static_pad_template_get_caps(staticPadTemplate));
+            unsigned length = gst_caps_get_size(caps.get());
+
+            GST_TRACE("factory %s caps has size %u", GST_OBJECT_NAME(factory), length);
+            for (unsigned i = 0; !decryptor && i < length; ++i) {
+                GstStructure* structure = gst_caps_get_structure(caps.get(), i);
+                GST_TRACE("checking structure %s", gst_structure_get_name(structure));
+                if (gst_structure_has_field_typed(structure, GST_PROTECTION_SYSTEM_ID_CAPS_FIELD, G_TYPE_STRING)) {
+                    const char* requestedProtectionSystemUuidUuid = gst_structure_get_string(structure, GST_PROTECTION_SYSTEM_ID_CAPS_FIELD);
```
Comment (calvaris) [inline Source/WebCore/platform/graphics/gstreamer/mse/AppendPipeline.cpp]: Too much search and replace.

`sysId` -> `protectionSystemUuid`
